{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "# 5. Applying QC cut-offs, annotating low quality variants and removing variants in LCRs\n",
    "\n",
    "For this stage, I used a mem2_ssd1_v2_x8 instance, using either 10 nodes (for smaller chromosomes) or 25 nodes (for larger chromosomes). I ran 1/2 the choromosomes all within one instance, and the other half within the other instance. The total cost for running this stage was Â£33. \n",
    "\n",
    "QC'd matrix tables are saved using DNAX. These tables total 318GB. \n",
    "\n",
    "## Set up environment\n",
    "Make sure you run this block only once. You'll get errors if you try to initialise Hail multiple times. If you do do this, you'll need to restart the kernel, and then initialise Hail only once. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": [],
    "trusted": true
   },
   "outputs": [],
   "source": [
    "# Initialise hail and spark logs? Running this cell will output a red-colored message- this is expected.\n",
    "# The 'Welcome to Hail' message in the output will indicate that Hail is ready to use in the notebook.\n",
    "import pyspark.sql\n",
    "\n",
    "config = pyspark.SparkConf().setAll([('spark.kryoserializer.buffer.max', '128')])\n",
    "sc = pyspark.SparkContext(conf=config) \n",
    "\n",
    "from pyspark.sql import SparkSession\n",
    "\n",
    "import hail as hl\n",
    "builder = (\n",
    "    SparkSession\n",
    "    .builder\n",
    "    .enableHiveSupport()\n",
    ")\n",
    "spark = builder.getOrCreate()\n",
    "hl.init(sc=sc)\n",
    "\n",
    "import dxpy"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The code below includes multiple stages, with heading matching the heading here: \n",
    "## Read in matrix table\n",
    "Here, you read in the matrix table which was saved out in step 1 from a DNAX database. First, set the chromosomes for which you are running this script. \n",
    "\n",
    "## Applying variant QC cut-offs and annotations\n",
    "### Filtering out very low quality variants, and annotating fairly low quality variants\n",
    "In script 2 you should have plotted variant QC metrics and from these plots, defined filters you wish to apply. Here we are removing variants: \n",
    "- With a GQ <= 30\n",
    "- With a call rate <= 0.7\n",
    "- Which lie within low complexity regions.\n",
    "\n",
    "And we are annotating variants as low quality if they have:\n",
    "- Mean GQ across samples <= 40\n",
    "- Call rate <= 0.9\n",
    "\n",
    "### Removing variants in LCRs\n",
    "Next, remove any variants lying within LCRs (low complexity regions) of the genome. \n",
    "\n",
    "## Filtering on sample QC metrics \n",
    "Remove samples with a low mean call rate \n",
    "\n",
    "## Saving out a QC'd matrix table \n",
    "Save out the post-variant and sample QC mtrix table using DNAX. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": [],
    "trusted": true
   },
   "outputs": [],
   "source": [
    "# Define the chromosomes you're working with (smaller half if working with fewer nodes, or larger half if working with more nodes)\n",
    "#chromosomes = list([4]+range(8,11)+ range(13,16)+[18]+range(20,24)) # For smaller chromosomes\n",
    "chromosomes = list(range(1,4)+ range(5,8)+range(11,13)+[17]+[19]) #For larger chromosomes \n",
    "\n",
    "for chr in chromosomes:\n",
    "    print(f'Processing chr {chr}...')\n",
    "\n",
    "\n",
    "\n",
    "    ######### Read in matrix table  ##########\n",
    "    print(f'Reading in MT for chr {chr}...')\n",
    "    mt=hl.read_matrix_table(f\"dnax://database-GgPbpq8J637bkp84VQyQ83X9/chromosome_{chr}_post_genoqc_final.mt\")\n",
    "    # Check this table is as you'd expect\n",
    "    print(f'Read in poost-genotype QC table from stage 1 for chr {chr}, sized {mt.count()}')\n",
    "    print(f'This table is formed of {mt.n_partitions()} partitions')\n",
    "    print(f'The structure of this MT is:')\n",
    "    print(mt.describe())\n",
    "    print(f'Finished reading in MT for chr {chr}!')\n",
    "\n",
    "\n",
    "\n",
    "    ####### Applying variant QC cut-offs and annotations ########\n",
    "    #### Filtering out very low quality variants, and annotating fairly low quality variants \n",
    "    print(f'Applying variant-level QC cut-offs for chr {chr}...')\n",
    "    # Filter out the lowest quality variants\n",
    "    print(f'n original variants in chromosome {chr}: {mt.count_rows()} following script 1')\n",
    "    mt=mt.filter_rows(mt.variant_qc.gq_stats.mean>30)\n",
    "    print(f'n variants in chromosome {chr} following removal of those with GQ<=30: {mt.count_rows()}')\n",
    "    mt=mt.filter_rows(mt.variant_qc.call_rate>0.7)\n",
    "    print(f'n variants in chromosome {chr} following removal of those with a call rate <=0.7: {mt.count_rows()}')\n",
    "    # Checkpoint here to speed up later stages of this script\n",
    "    mt.checkpoint('post_lq_vars.mt', overwrite=True)\n",
    "    # Annotate variants which are fairly low quality so these can be removed at a later point. \n",
    "    mt = mt.annotate_rows(variant_gq_over_40 = mt.variant_qc.gq_stats.mean >40)\n",
    "    variant_gq_over_40_count = mt.aggregate_rows(hl.agg.count_where(mt.variant_gq_over_40))\n",
    "    print(f'n variants in chromosome {chr} with variant GQ > 40: {variant_gq_over_40_count}')\n",
    "    mt = mt.annotate_rows(variant_call_rate_over_90_percent = mt.variant_qc.call_rate > 0.9)\n",
    "    variant_call_rate_over_90_percent_count = mt.aggregate_rows(hl.agg.count_where(mt.variant_call_rate_over_90_percent))\n",
    "    print(f'n variants in chromosome {chr} with call rate > 0.9: {variant_call_rate_over_90_percent_count}')\n",
    "    HQ_var_count=mt.aggregate_rows(hl.agg.count_where(mt.variant_call_rate_over_90_percent & mt.variant_gq_over_40))\n",
    "    print(f'n high quality variants in chromosome {chr}: {HQ_var_count}')\n",
    "\n",
    "    #### Removing variants in LCRs\n",
    "    lcr_url = f\"file:///mnt/project/LCR_intervals.tsv\"\n",
    "    intervals=hl.import_locus_intervals(lcr_url, reference_genome='GRCh38', skip_invalid_intervals=True)\n",
    "    mt=mt.annotate_rows(LCR=hl.is_defined(intervals[mt.locus]))\n",
    "    mt=mt.filter_rows(mt.LCR==False)    \n",
    "    print(f'n high quality variants in chromosome {chr} following removal of those in LCRs: {mt.count()}')\n",
    "    # Checkpoint again to speed up later stages of the script\n",
    "    mt.checkpoint('post_lcr_rm.mt', overwrite=True)\n",
    "    print(f'Finished variant-level QC filtering for chr {chr}!')\n",
    "\n",
    "\n",
    "    ###### Filtering on sample QC metrics #######\n",
    "    print(f'Applying sample-level QC cut-offs for chr {chr}...')\n",
    "    print(f'Samples before any QC applied: {mt.count_cols()}')\n",
    "    ids_passing_QC = f\"file:///mnt/project/WES_QC/ids_with_call_rate_over_80_percent.tsv\"\n",
    "    samples_to_keep=hl.import_table(ids_passing_QC)\n",
    "    samples_to_keep=samples_to_keep.rename({'\"x\"':\"s\"})\n",
    "    samples_to_keep=samples_to_keep.key_by(samples_to_keep.s)\n",
    "    mt = mt.filter_cols((hl.is_defined(samples_to_keep[mt.s])))\n",
    "    print(f'Samples remaining following removal of those with call rate < 0.8: {mt.count_cols()}')\n",
    "    print(f'Finished sample-level QC filtering for chr {chr}!')\n",
    "\n",
    "    ###### Saving out a QC'd matrix table \n",
    "    print(f'Saving out QCd MT for chr {chr}...')\n",
    "    # Define database and MT names\n",
    "    # Note: It is recommended to only use lowercase letters for the database name.\n",
    "    # # If uppercase lettering is used, the database name will be lowercased when creating the database.\n",
    "    db_name = f\"ukbb_test_wes_matrix_tables_post_qc_filtering\"\n",
    "    mt_name = f\"chromosome_{chr}_post_geno_sample_and_var_qc.mt\"\n",
    "    # Create database in DNAX\n",
    "    stmt = f\"CREATE DATABASE IF NOT EXISTS {db_name} LOCATION 'dnax://'\"\n",
    "    print(stmt)\n",
    "    spark.sql(stmt).show()\n",
    "    # Store MT in DNAX\n",
    "    # # Find database ID of newly created database using dxpy method\n",
    "    db_uri = dxpy.find_one_data_object(name=f\"{db_name}\", classname=\"database\")['id']\n",
    "    url = f\"dnax://{db_uri}/{mt_name}\" # Note: the dnax url must follow this format to properly save MT to DNAX\n",
    "    \n",
    "    # Before this step, the Hail MatrixTable is just an object in memory. To persist it and be able to access it later, the notebook needs to write it into a persistent filesystem (in this case DNAX).\n",
    "    mt.write(url) # Note: output should describe size of MT (i.e. number of rows, columns, partitions) \n",
    "    print(f'QCd MT for chr {chr} saved! Now checking file can be read in ... ')\n",
    "    \n",
    "    \n",
    "    ###### Check your mt can be read in for later scripts\n",
    "    b=hl.read_matrix_table(f\"dnax://database-Ggy1X3QJ637qBKGypjy9y9f4/chromosome_{chr}_post_geno_sample_and_var_qc.mt\")\n",
    "    # Check count is as you'd expect\n",
    "    print(f'Count for read in matrix table is {b.count()} - check this matches what you expect!')\n",
    "    # Check the additional fields you added are present\n",
    "    print(f'The structure of the post-QC MT is:')\n",
    "    print(b.describe())\n",
    "    \n",
    "    \n",
    "    print(f'This stage of processing for chr {chr} complete!')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
