{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 9. Applying QC cut-offs and VEP annotations\n",
    "\n",
    "This script reads in matrix tables, applies QC filters and adds VEP annotations so these tables are then ready for variants to be counted. \n",
    "\n",
    "For this stage I used a mem2_ssd1_v2_x8 instance with 10 nodes, and used this one instance to run over all chromosomes. The total cost for this stage was Â£13. QC'd matrix tables are saved out using DNAX. \n",
    "\n",
    "## Set up environment\n",
    "Make sure you run this block only once. You'll get errors if you try to initialise Hail multiple times. If you do do this, you'll need to restart the kernel, and then initialise Hail only once. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initialise hail and spark logs? Running this cell will output a red-colored message- this is expected.\n",
    "# The 'Welcome to Hail' message in the output will indicate that Hail is ready to use in the notebook.\n",
    "import pyspark.sql\n",
    "\n",
    "config = pyspark.SparkConf().setAll([('spark.kryoserializer.buffer.max', '128')])\n",
    "sc = pyspark.SparkContext(conf=config) \n",
    "\n",
    "from pyspark.sql import SparkSession\n",
    "\n",
    "import hail as hl\n",
    "builder = (\n",
    "    SparkSession\n",
    "    .builder\n",
    "    .enableHiveSupport()\n",
    ")\n",
    "spark = builder.getOrCreate()\n",
    "hl.init(sc=sc)\n",
    "\n",
    "import dxpy"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read in files needed for annotations across all chromosomes \n",
    "Here, read in (and key correctly for annotation): \n",
    "- the IDs of unrelated samples you want to keep (from stage 6). \n",
    "- REVEL scores \n",
    "- GnomAD annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# UNRELATED IDS\n",
    "### Read in unrealted sample IDs to keep in (this file was created in stage 06).\n",
    "samples_to_keep=hl.import_table(f\"file:///mnt/project/WES_QC/3rd_degree_rel_ids_to_keep_500k.csv\", delimiter=',')\n",
    "samples_to_keep.count()\n",
    "samples_to_keep=samples_to_keep.rename({'\"x\"':\"s\"})\n",
    "samples_to_keep=samples_to_keep.key_by(samples_to_keep.s)\n",
    "\n",
    "# REVEL SCORES \n",
    "# Read in REVEL file, defining the locus field as type = locus so it can be annotated to other loci\n",
    "revel=hl.import_table(f\"file:///mnt/project/matching_200k_analysis/revel_scores.tsv\", \n",
    "                      delimiter= '\\t',\n",
    "                      types={'locus': hl.expr.types.tlocus(reference_genome='GRCh38'), 'alt': hl.tstr, 'REVEL_score': hl.tfloat64}) \n",
    "print(revel.count())\n",
    "revel=revel.key_by(revel.locus, revel.alt)\n",
    "\n",
    "# GNOMAD\n",
    "# Read in gnomad file \n",
    "## In terminal read in file from project using 'dx download matching_200k_analysis/constraint_scores.tsv' \n",
    "constraint=hl.import_table(f\"file:///mnt/project/matching_200k_analysis/constraint_scores.tsv\", \n",
    "                           delimiter='\\t',\n",
    "                           types={'pLI': hl.tfloat64, 'oe_lof':hl.tfloat64, 'oe_lof_upper':hl.tfloat64})\n",
    "constraint.count()\n",
    "constraint=constraint.key_by('gene_id') # This is the ensembl id, so like ENSG000..... \n",
    "# Rename to match matrix table \n",
    "constraint=constraint.rename({\"gene_id\":\"gene_id_worstCsq\"})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read in matrix tables per chromosome, filter, and annotate \n",
    "This whole process is within a loop that happens per chromosome. Note that chromosome 1 is treated differently. Chromosome 1 was too big to effectively work with as one matrix table, so here it is split into two halves and then each half is worked with individually from this point. \n",
    "\n",
    "\n",
    "Each section of the loop is described below: \n",
    "- Reading in matrix table: Here, you read in the matrix table which was saved out in stage 5 from a DNAX database. First, set the chromosome for which you are running this script. When working with chromosome 1, the mt is split into 2 halves at this point.  \n",
    "\n",
    "- Filter to unrelated samples: This section uses the IDs of the unrelated samples you read in above to filter the matrix table so it only contains unrealted samples.\n",
    "\n",
    "- Add VEP annotations: This sections uses the VEP annotations saved out in stage 4 to annotate the matrix table with variant consequences. The key of these tables matches so this is a simple annotation to add. Then set up PTV, NS, miss and synonymous annotations and worst consequences, and save out the matrix table to reduce computational load for next steps. \n",
    "\n",
    "- Add REVEL annotations: This section uses the REVEL annotations you read in above to annotate the matrix table with REVEL scores. This table is keyed by locus and alternate allele, so we need to form an alternate allele column within our matrix table, and then key by it. Once the keys are the same, the annotation can be added.\n",
    "\n",
    "- Add per gene constraint metrics: This section uses the GnomAD annotations you read in above to annotate the matrix table with LOEUF and pLI scores per gene. This requires keying by gene ID. Here, we also rekey the table so it is keyed as it originally was and then write out the matrix table and read it back in to speed up processing. \n",
    "\n",
    "- Variant QC: filter MT to remove low quality variants highlighted in stage 5. \n",
    "\n",
    "- Allele counts: Run variant_qc command in hail to redo allele counts in this high quality sample and variant matrix table. These counts will later be used to filter to rare variants. \n",
    "\n",
    "- Write out: Write the annotated and QC'd matrix table out to DNAX. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Filter and annotate MTs \n",
    "chromosomes=list(range(1,23))\n",
    "\n",
    "for chr in chromosomes:\n",
    "    if chr == 1:\n",
    "        print(f\"Processing chromosome {chr} ...\")\n",
    "\n",
    "        ## First half\n",
    "        ####### READ IN MT #########\n",
    "        mt=hl.read_matrix_table(f\"dnax://database-Ggy1X3QJ637qBKGypjy9y9f4/chromosome_{chr}_post_geno_sample_and_var_qc.mt\")\n",
    "        # Check this table is as you'd expect\n",
    "        print(f\"Size when initially read in: {mt.count()}\")\n",
    "        # Split into two halves \n",
    "        total_rows=mt.count_rows()\n",
    "        split_index=total_rows//2\n",
    "        print(split_index)\n",
    "        mt=mt.add_row_index()\n",
    "        # First half\n",
    "        mt_first_half=mt.filter_rows(hl.int(mt.row_idx)<split_index)\n",
    "        print(mt_first_half.count())\n",
    "        # Check you're happy with this split (that it follows on from each bit correctly)\n",
    "        mt_first_half.tail(10)\n",
    "        mt_first_half.checkpoint(f'chr{chr}_1sthalf.mt', overwrite=True)\n",
    "        mt=hl.read_matrix_table(f'chr{chr}_1sthalf.mt')\n",
    "\n",
    "        \n",
    "        ####### REMOVE RELATED SAMPLES ######\n",
    "        mt = mt.filter_cols((hl.is_defined(samples_to_keep[mt.s])))\n",
    "        print(f'MT size after removing related samples: {mt.count()}')\n",
    "\n",
    "\n",
    "        ####### ADD VEP ANNOTATIONS #########\n",
    "        # Read in VEP annotations file \n",
    "        annotations = hl.read_table(f\"dnax://database-Ggy5qPjJ637zF5j5yyX9bGxG/chr_{chr}_annotations_LoFTEE_updated.ht\")\n",
    "        print(f'Total annotations file size (should be bigger than mt): {annotations.count()}') # This should be bigger than n variants you have, as its from an unQCd version of this file\n",
    "        # Set up annotations\n",
    "        PTV_annotations = hl.set([\"splice_acceptor_variant\", \"splice_donor_variant\", \"stop_gained\", \"frameshift_variant\"])\n",
    "        NS_annotations = hl.set([\"splice_acceptor_variant\", \"splice_donor_variant\", \"stop_gained\", \"inframe_insertion\", \"inframe_deletion\", \"inframe_insertion\", \"missense_variant\", \"stop_lost\", \"start_lost\", \"frameshift_variant\"])\n",
    "        Miss_annotations = hl.set([\"missense_variant\"])\n",
    "        S_annotations = hl.set([\"synonymous_variant\"])\n",
    "        # Annotate\n",
    "        annotated_mt = annotated_mt.annotate_rows(LoF_worstCsq = (PTV_annotations.contains(annotated_mt.vep.most_severe_consequence)),\n",
    "                          NS_worstCsq = (NS_annotations.contains(annotated_mt.vep.most_severe_consequence)),\n",
    "                          Miss_worstCsq = (Miss_annotations.contains(annotated_mt.vep.most_severe_consequence)),\n",
    "                          Syn_worstCsq = (S_annotations.contains(annotated_mt.vep.most_severe_consequence)),\n",
    "                          gene_symbol_worstCsq = (annotated_mt.vep.transcript_consequences.find(lambda x : x.consequence_terms.contains(annotated_mt.vep.most_severe_consequence)).gene_symbol),\n",
    "                          gene_id_worstCsq = (annotated_mt.vep.transcript_consequences.find(lambda x : x.consequence_terms.contains(annotated_mt.vep.most_severe_consequence)).gene_id)\n",
    "                         )\n",
    "        print(f'Structure of VEP annotated MT (check this has all annotations you want!): {annotated_mt.describe()}') # Check these annotations have all been added\n",
    "\n",
    "\n",
    "        ######## ADD REVEL ANNOTATIONS #######\n",
    "        s=annotated_mt.alleles[1]\n",
    "        annotated_mt=annotated_mt.annotate_rows(alt=s)\n",
    "        # Rekey annotated mt\n",
    "        annotated_mt=annotated_mt.key_rows_by(annotated_mt.locus, annotated_mt.alt)\n",
    "        # Annotate\n",
    "        vep_revel_mt=annotated_mt.annotate_rows(REVEL_score=revel[annotated_mt.locus, annotated_mt.alt].REVEL_score)\n",
    "        print(f'Structure of REVEL annotated MT (check this has all annotations you want!): {vep_revel_mt.describe()}') # Check these annotations have all been added\n",
    "\n",
    "\n",
    "        ####### ADD GNOMAD ANNOTATIONS ########\n",
    "        #Rekey to match \n",
    "        vep_revel_mt=vep_revel_mt.key_rows_by(vep_revel_mt.gene_id_worstCsq)\n",
    "        # Annotate \n",
    "        vep_revel_constraint_mt=vep_revel_mt.annotate_rows(LoF_gene_pLI_score =constraint[vep_revel_mt.gene_id_worstCsq].pLI)\n",
    "        vep_revel_constraint_mt=vep_revel_constraint_mt.annotate_rows(LoF_gene_pLI_LoFi =constraint[vep_revel_constraint_mt.gene_id_worstCsq].pli_9)\n",
    "        vep_revel_constraint_mt=vep_revel_constraint_mt.annotate_rows(LoF_gene_oe_lof_upper =constraint[vep_revel_constraint_mt.gene_id_worstCsq].oe_lof_upper)\n",
    "        vep_revel_constraint_mt=vep_revel_constraint_mt.annotate_rows(LoF_gene_LOEUF_LoFi =constraint[vep_revel_constraint_mt.gene_id_worstCsq].loeuf_35)\n",
    "        print(f'Structure of constraint annotated MT (check this has all annotations you want!): {vep_revel_constraint_mt.describe()}') # Check these annotations have all been added\n",
    "        # Checkpoint here to speed up processing\n",
    "        vep_revel_constraint_mt.checkpoint(f'vep_annotated_mt_chr{chr}_1sthalf.mt', overwrite=True)\n",
    "        mt=hl.read_matrix_table(f'vep_annotated_mt_chr{chr}_1sthalf.mt')\n",
    "\n",
    "\n",
    "        ####### VARIANT QC #########\n",
    "        # Remove variants based on final QC stages\n",
    "        print(f'n variants in chromosome {chr} first half:{mt.count_rows()}') \n",
    "        mt=mt.filter_rows(mt.variant_gq_over_40==True)\n",
    "        print(f'n variants in chromosome {chr} first half with variant GQ > 40: {mt.count_rows()}')\n",
    "        mt=mt.filter_rows(mt.variant_call_rate_over_90_percent==True)\n",
    "        print(f'n variants in chromosome {chr} first half with call rate > 0.9: {mt.count_rows()}')\n",
    "\n",
    "\n",
    "        ####### ALLELE COUNTS #####\n",
    "        # Rerun variant QC command so allele counts you work with are post all QC \n",
    "        mt = hl.variant_qc(mt)\n",
    "        print(f'Variant QC run and allele counts updated for chr {chr} first half. MT sized {mt.count()}')\n",
    "    \n",
    "\n",
    "        ####### SAVE OUT MT ########\n",
    "        #Write out matrix table here so it can be read in from here for future counts!\n",
    "        db_name = f\"mt_count_ready\"\n",
    "        mt_name = f\"chr_{chr}_first_half_ready_for_counts.mt\"\n",
    "        # Create database in DNAX\n",
    "        stmt = f\"CREATE DATABASE IF NOT EXISTS {db_name} LOCATION 'dnax://'\"\n",
    "        print(stmt)\n",
    "        spark.sql(stmt).show()\n",
    "        # Store MT in DNAX\n",
    "        import dxpy\n",
    "        # Find database ID of newly created database using dxpy method\n",
    "        db_uri = dxpy.find_one_data_object(name=f\"{db_name}\", classname=\"database\")['id']\n",
    "        url = f\"dnax://{db_uri}/{mt_name}\"\n",
    "        mt.checkpoint(url) \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "        ## Second half\n",
    "        ######### READ IN MT ########## \n",
    "        mt=hl.read_matrix_table(f\"dnax://database-Ggy1X3QJ637qBKGypjy9y9f4/chromosome_{chr}_post_geno_sample_and_var_qc.mt\")\n",
    "        # Check this table is as you'd expect\n",
    "        print(f\"Size when initially read in: {mt.count()}\")\n",
    "        # Split into two halves \n",
    "        total_rows=mt.count_rows()\n",
    "        split_index=total_rows//2\n",
    "        print(split_index)\n",
    "        mt=mt.add_row_index()\n",
    "        # Second half\n",
    "        mt_second_half=mt.filter_rows(hl.int(mt.row_idx)>=split_index) \n",
    "        print(mt_second_half.count())\n",
    "        # Check you're happy with this split (that it follows on from each bit correctly)\n",
    "        mt_second_half.head(10)\n",
    "        mt_first_half.checkpoint(f'chr{chr}_2ndhalf.mt', overwrite=True)\n",
    "        mt=hl.read_matrix_table(f'chr{chr}_2ndhalf.mt')\n",
    "\n",
    "        \n",
    "        ####### REMOVE RELATED SAMPLES ######\n",
    "        mt = mt.filter_cols((hl.is_defined(samples_to_keep[mt.s])))\n",
    "        print(f'MT size after removing related samples: {mt.count()}')\n",
    "\n",
    "\n",
    "        ####### ADD VEP ANNOTATIONS #########\n",
    "        # Read in VEP annotations file \n",
    "        annotations = hl.read_table(f\"dnax://database-Ggy5qPjJ637zF5j5yyX9bGxG/chr_{chr}_annotations_LoFTEE_updated.ht\")\n",
    "        print(f'Total annotations file size (should be bigger than mt): {annotations.count()}') # This should be bigger than n variants you have, as its from an unQCd version of this file\n",
    "        # Set up annotations\n",
    "        PTV_annotations = hl.set([\"splice_acceptor_variant\", \"splice_donor_variant\", \"stop_gained\", \"frameshift_variant\"])\n",
    "        NS_annotations = hl.set([\"splice_acceptor_variant\", \"splice_donor_variant\", \"stop_gained\", \"inframe_insertion\", \"inframe_deletion\", \"inframe_insertion\", \"missense_variant\", \"stop_lost\", \"start_lost\", \"frameshift_variant\"])\n",
    "        Miss_annotations = hl.set([\"missense_variant\"])\n",
    "        S_annotations = hl.set([\"synonymous_variant\"])\n",
    "        # Annotate\n",
    "        annotated_mt = annotated_mt.annotate_rows(LoF_worstCsq = (PTV_annotations.contains(annotated_mt.vep.most_severe_consequence)),\n",
    "                          NS_worstCsq = (NS_annotations.contains(annotated_mt.vep.most_severe_consequence)),\n",
    "                          Miss_worstCsq = (Miss_annotations.contains(annotated_mt.vep.most_severe_consequence)),\n",
    "                          Syn_worstCsq = (S_annotations.contains(annotated_mt.vep.most_severe_consequence)),\n",
    "                          gene_symbol_worstCsq = (annotated_mt.vep.transcript_consequences.find(lambda x : x.consequence_terms.contains(annotated_mt.vep.most_severe_consequence)).gene_symbol),\n",
    "                          gene_id_worstCsq = (annotated_mt.vep.transcript_consequences.find(lambda x : x.consequence_terms.contains(annotated_mt.vep.most_severe_consequence)).gene_id)\n",
    "                         )\n",
    "        print(f'Structure of VEP annotated MT (check this has all annotations you want!): {annotated_mt.describe()}') # Check these annotations have all been added\n",
    "\n",
    "\n",
    "        ######## ADD REVEL ANNOTATIONS #######\n",
    "        s=annotated_mt.alleles[1]\n",
    "        annotated_mt=annotated_mt.annotate_rows(alt=s)\n",
    "        # Rekey annotated mt\n",
    "        annotated_mt=annotated_mt.key_rows_by(annotated_mt.locus, annotated_mt.alt)\n",
    "        # Annotate\n",
    "        vep_revel_mt=annotated_mt.annotate_rows(REVEL_score=revel[annotated_mt.locus, annotated_mt.alt].REVEL_score)\n",
    "        print(f'Structure of REVEL annotated MT (check this has all annotations you want!): {vep_revel_mt.describe()}') # Check these annotations have all been added\n",
    "\n",
    "\n",
    "        ####### ADD GNOMAD ANNOTATIONS ########\n",
    "        #Rekey to match \n",
    "        vep_revel_mt=vep_revel_mt.key_rows_by(vep_revel_mt.gene_id_worstCsq)\n",
    "        # Annotate \n",
    "        vep_revel_constraint_mt=vep_revel_mt.annotate_rows(LoF_gene_pLI_score =constraint[vep_revel_mt.gene_id_worstCsq].pLI)\n",
    "        vep_revel_constraint_mt=vep_revel_constraint_mt.annotate_rows(LoF_gene_pLI_LoFi =constraint[vep_revel_constraint_mt.gene_id_worstCsq].pli_9)\n",
    "        vep_revel_constraint_mt=vep_revel_constraint_mt.annotate_rows(LoF_gene_oe_lof_upper =constraint[vep_revel_constraint_mt.gene_id_worstCsq].oe_lof_upper)\n",
    "        vep_revel_constraint_mt=vep_revel_constraint_mt.annotate_rows(LoF_gene_LOEUF_LoFi =constraint[vep_revel_constraint_mt.gene_id_worstCsq].loeuf_35)\n",
    "        print(f'Structure of constraint annotated MT (check this has all annotations you want!): {vep_revel_constraint_mt.describe()}') # Check these annotations have all been added\n",
    "        # Checkpoint here to speed up processing\n",
    "        vep_revel_constraint_mt.checkpoint(f'vep_annotated_mt_chr{chr}_2ndhalf.mt', overwrite=True)\n",
    "        mt=hl.read_matrix_table(f'vep_annotated_mt_chr{chr}_2ndhalf.mt')\n",
    "\n",
    "\n",
    "        ####### VARIANT QC #########\n",
    "        # Remove variants based on final QC stages\n",
    "        print(f'n variants in chromosome {chr} second half:{mt.count_rows()}') \n",
    "        mt=mt.filter_rows(mt.variant_gq_over_40==True)\n",
    "        print(f'n variants in chromosome {chr} second half with variant GQ > 40: {mt.count_rows()}')\n",
    "        mt=mt.filter_rows(mt.variant_call_rate_over_90_percent==True)\n",
    "        print(f'n variants in chromosome {chr} second half with call rate > 0.9: {mt.count_rows()}')\n",
    "\n",
    "\n",
    "        ####### ALLELE COUNTS #####\n",
    "        # Rerun variant QC command so allele counts you work with are post all QC \n",
    "        mt = hl.variant_qc(mt)\n",
    "        print(f'Variant QC run and allele counts updated for chr {chr} second half. MT sized {mt.count()}')\n",
    "    \n",
    "\n",
    "        ####### SAVE OUT MT ########\n",
    "        #Write out matrix table here so it can be read in from here for future counts!\n",
    "        db_name = f\"mt_count_ready\"\n",
    "        mt_name = f\"chr_{chr}_second_half_ready_for_counts.mt\"\n",
    "        # Create database in DNAX\n",
    "        stmt = f\"CREATE DATABASE IF NOT EXISTS {db_name} LOCATION 'dnax://'\"\n",
    "        print(stmt)\n",
    "        spark.sql(stmt).show()\n",
    "        # Store MT in DNAX\n",
    "        import dxpy\n",
    "        # Find database ID of newly created database using dxpy method\n",
    "        db_uri = dxpy.find_one_data_object(name=f\"{db_name}\", classname=\"database\")['id']\n",
    "        url = f\"dnax://{db_uri}/{mt_name}\"\n",
    "        mt.checkpoint(url) \n",
    "\n",
    "\n",
    "\n",
    "        print(f'Chromosome {chr} annotated and matrix tables (per half) written out!')\n",
    "    else:\n",
    "        print(f\"Processing chromosome {chr} ...\")\n",
    "\n",
    "        ####### READ IN MT #########\n",
    "        mt=hl.read_matrix_table(f\"dnax://database-Ggy1X3QJ637qBKGypjy9y9f4/chromosome_{chr}_post_geno_sample_and_var_qc.mt\")\n",
    "        # Check this table is as you'd expect\n",
    "        print(f\"Size when initially read in: {mt.count()}\")\n",
    "\n",
    "        ####### REMOVE RELATED SAMPLES ######\n",
    "        mt = mt.filter_cols((hl.is_defined(samples_to_keep[mt.s])))\n",
    "        print(f'MT size after removing related samples: {mt.count()}')\n",
    "\n",
    "\n",
    "        ####### ADD VEP ANNOTATIONS #########\n",
    "        # Read in VEP annotations file \n",
    "        annotations = hl.read_table(f\"dnax://database-Ggy5qPjJ637zF5j5yyX9bGxG/chr_{chr}_annotations_LoFTEE_updated.ht\")\n",
    "        print(f'Total annotations file size (should be bigger than mt): {annotations.count()}') # This should be bigger than n variants you have, as its from an unQCd version of this file\n",
    "        # Set up annotations\n",
    "        PTV_annotations = hl.set([\"splice_acceptor_variant\", \"splice_donor_variant\", \"stop_gained\", \"frameshift_variant\"])\n",
    "        NS_annotations = hl.set([\"splice_acceptor_variant\", \"splice_donor_variant\", \"stop_gained\", \"inframe_insertion\", \"inframe_deletion\", \"inframe_insertion\", \"missense_variant\", \"stop_lost\", \"start_lost\", \"frameshift_variant\"])\n",
    "        Miss_annotations = hl.set([\"missense_variant\"])\n",
    "        S_annotations = hl.set([\"synonymous_variant\"])\n",
    "        # Annotate\n",
    "        annotated_mt = annotated_mt.annotate_rows(LoF_worstCsq = (PTV_annotations.contains(annotated_mt.vep.most_severe_consequence)),\n",
    "                          NS_worstCsq = (NS_annotations.contains(annotated_mt.vep.most_severe_consequence)),\n",
    "                          Miss_worstCsq = (Miss_annotations.contains(annotated_mt.vep.most_severe_consequence)),\n",
    "                          Syn_worstCsq = (S_annotations.contains(annotated_mt.vep.most_severe_consequence)),\n",
    "                          gene_symbol_worstCsq = (annotated_mt.vep.transcript_consequences.find(lambda x : x.consequence_terms.contains(annotated_mt.vep.most_severe_consequence)).gene_symbol),\n",
    "                          gene_id_worstCsq = (annotated_mt.vep.transcript_consequences.find(lambda x : x.consequence_terms.contains(annotated_mt.vep.most_severe_consequence)).gene_id)\n",
    "                         )\n",
    "        print(f'Structure of VEP annotated MT (check this has all annotations you want!): {annotated_mt.describe()}') # Check these annotations have all been added\n",
    "\n",
    "\n",
    "        ######## ADD REVEL ANNOTATIONS #######\n",
    "        s=annotated_mt.alleles[1]\n",
    "        annotated_mt=annotated_mt.annotate_rows(alt=s)\n",
    "        # Rekey annotated mt\n",
    "        annotated_mt=annotated_mt.key_rows_by(annotated_mt.locus, annotated_mt.alt)\n",
    "        # Annotate\n",
    "        vep_revel_mt=annotated_mt.annotate_rows(REVEL_score=revel[annotated_mt.locus, annotated_mt.alt].REVEL_score)\n",
    "        print(f'Structure of REVEL annotated MT (check this has all annotations you want!): {vep_revel_mt.describe()}') # Check these annotations have all been added\n",
    "\n",
    "\n",
    "        ####### ADD GNOMAD ANNOTATIONS ########\n",
    "        #Rekey to match \n",
    "        vep_revel_mt=vep_revel_mt.key_rows_by(vep_revel_mt.gene_id_worstCsq)\n",
    "        # Annotate \n",
    "        vep_revel_constraint_mt=vep_revel_mt.annotate_rows(LoF_gene_pLI_score =constraint[vep_revel_mt.gene_id_worstCsq].pLI)\n",
    "        vep_revel_constraint_mt=vep_revel_constraint_mt.annotate_rows(LoF_gene_pLI_LoFi =constraint[vep_revel_constraint_mt.gene_id_worstCsq].pli_9)\n",
    "        vep_revel_constraint_mt=vep_revel_constraint_mt.annotate_rows(LoF_gene_oe_lof_upper =constraint[vep_revel_constraint_mt.gene_id_worstCsq].oe_lof_upper)\n",
    "        vep_revel_constraint_mt=vep_revel_constraint_mt.annotate_rows(LoF_gene_LOEUF_LoFi =constraint[vep_revel_constraint_mt.gene_id_worstCsq].loeuf_35)\n",
    "        print(f'Structure of constraint annotated MT (check this has all annotations you want!): {vep_revel_constraint_mt.describe()}') # Check these annotations have all been added\n",
    "        # Checkpoint here to speed up processing\n",
    "        vep_revel_constraint_mt.checkpoint(f'vep_annotated_mt_chr{chr}.mt', overwrite=True)\n",
    "        mt=hl.read_matrix_table(f'vep_annotated_mt_chr{chr}.mt')\n",
    "\n",
    "\n",
    "        ####### VARIANT QC #########\n",
    "        # Remove variants based on final QC stages\n",
    "        print(f'n variants in chromosome {chr}:{mt.count_rows()}') \n",
    "        mt=mt.filter_rows(mt.variant_gq_over_40==True)\n",
    "        print(f'n variants in chromosome {chr} with variant GQ > 40: {mt.count_rows()}')\n",
    "        mt=mt.filter_rows(mt.variant_call_rate_over_90_percent==True)\n",
    "        print(f'n variants in chromosome {chr} with call rate > 0.9: {mt.count_rows()}')\n",
    "\n",
    "\n",
    "        ####### ALLELE COUNTS #####\n",
    "        # Rerun variant QC command so allele counts you work with are post all QC \n",
    "        mt = hl.variant_qc(mt)\n",
    "        print(f'Variant QC run and allele counts updated for chr {chr}. MT sized {mt.count()}')\n",
    "    \n",
    "\n",
    "        ####### SAVE OUT MT ########\n",
    "        #Write out matrix table here so it can be read in from here for future counts!\n",
    "        db_name = f\"mt_count_ready\"\n",
    "        mt_name = f\"chr_{chr}_ready_for_counts.mt\"\n",
    "        # Create database in DNAX\n",
    "        stmt = f\"CREATE DATABASE IF NOT EXISTS {db_name} LOCATION 'dnax://'\"\n",
    "        print(stmt)\n",
    "        spark.sql(stmt).show()\n",
    "        # Store MT in DNAX\n",
    "        import dxpy\n",
    "        # Find database ID of newly created database using dxpy method\n",
    "        db_uri = dxpy.find_one_data_object(name=f\"{db_name}\", classname=\"database\")['id']\n",
    "        url = f\"dnax://{db_uri}/{mt_name}\"\n",
    "        mt.checkpoint(url) \n",
    "\n",
    "\n",
    "        print(f'Chromosome {chr} annotated and matrix table written out!')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "name": "python",
   "version": "3.12.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
